h2o-core/src/main/java/water/Value.java

```diff
@@ -485,11 +492,18 @@ boolean setReplica( H2ONode h2o ) {
     // up, but the replica list does not account for the new replica.  However,
     // the rwlock cannot go down until an ACKACK is received, and the ACK
     // (hence ACKACK) doesn't go out until after this function returns.
-    replicas()[h2o._unique_idx] = 1;
+    markHotReplica(h2o);
     // Both rwlock taken, and replica count is up now.
     return true;
   }
 
+  private void markHotReplica(H2ONode n) {
+    n.markLocalDKVAccess();
+    byte[] r = replicas();
+    if (n._unique_idx < r.length)
```

Comment:
If I understand correctly, this is the main fix where we don't set replica to active on unseen nodes, is that correct?

Reply:
The fix does 2 things:
- instead of considering `CLOUD.size()` to make the replica array, it uses `H2ONode.IDX.length` - this is the correct way because any node that tries to communicate with H2O during cloud formation will be recorded in H2ONode.IDX array.
- we don't grow the array, instead we assume that in a regular deployment there will be just one client (but we keep buffer for 10), if the assumption is not valid we do not remember the replicas for the rest of the clients. Instead of invalidating per-value, we always invalidate regardless whether the node actually touched the value or not. We just remember whether the node touched the DKV at all instead.